File: algo/114. Flatten Binary Tree to Linked List.js
// Divide and Conquer

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */

var helper = function(root) {
    var head, bottom;
    if (root.left == null && root.right == null) {
        head = root;
        bottom = root;
    } else if (root.left != null && root.right != null) {
        var leftRes = helper(root.left);
        var rightRes = helper(root.right);
        root.left = null;
        root.right = leftRes[0];
        leftRes[1].right = rightRes[0];
        head = root;
        bottom = rightRes[1];
    } else if (root.left == null && root.right != null) {
        var rightRes = helper(root.right);
        head = root;
        bottom = rightRes[1];
    } else {
        var leftRes = helper(root.left);
        head = root;
        root.left = null;
        root.right = leftRes[0];
        bottom = leftRes[1];
    }
    
    return [head, bottom];
}

var flatten = function(root) {
    if (root == null) {
        return null;
    }
    
    return helper(root)[0];
};

File: round3/114. Flatten Binary Tree to Linked List.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public void Flatten(TreeNode root) {
        Helper(root);
    }
    
    // returns bottom right leaf
    public TreeNode Helper(TreeNode root) {
        if (root == null) return null;
        // leaf
        if (root.left == null && root.right == null) {
            return root;
        }
        
        // divide
        TreeNode leftRes = Helper(root.left);
        TreeNode rightRes = Helper(root.right);
        
        // merge        
        if (leftRes == null) return rightRes;
        leftRes.right = root.right;
        root.right = root.left;
        root.left = null; // very important to null the left child pointer
        
        return rightRes == null ? leftRes : rightRes;
    }
}